### PR TITLE
2.0.0-IDEA: Make the "as" transport header mandatory

### PR DESCRIPTION
**Pull request by Raynos**
This is another breaking change. All outgoing requests must
set the "as" header on the way out.

I also fixed a seperate bug in connection about making
assumptions on the `err` object.

I added a bonus commit to reject incoming call requests without
the "as" transport header; however thats a binary breaking change
so I probably want to roll it back as we have no agreement or
concensus on it and it requires everyone in the team to agree.

cc @jcorbin @kriskowal

### CODE REVIEW

File: node/channel.js [hunk 1]
```diff
@@ -83,6 +83,8 @@ function TChannel(options) {
     self.requestDefaults = extend({
         timeout: TChannelRequest.defaultTimeout
     }, self.options.requestDefaults);
+    self.transportHeaderDefaults =
```
Comment (jcorbin) [inline node/channel.js]: erm, how was this not already afforded for ala requestDefaults.headers?

File: node/channel.js [hunk 2]
```diff
@@ -356,11 +358,26 @@ TChannel.prototype.requestOptions = function requestOptions(options) {
     for (prop in self.requestDefaults) {
         opts[prop] = self.requestDefaults[prop];
     }
+    if (self.transportHeaderDefaults) {
+        opts.headers = {};
```
Comment (jcorbin) [inline node/channel.js]: let's just always set this empty, and fill it in...

File: node/channel.js [hunk 3]
```diff
@@ -356,11 +358,26 @@ TChannel.prototype.requestOptions = function requestOptions(options) {
...
     if (options) {
         for (prop in options) {
+            if (prop === 'headers') {
+                continue;
+            }
             opts[prop] = options[prop];
         }
     }
+    if (options && options.headers) {
+        opts.headers = opts.headers || {};
```
Comment (jcorbin) [inline node/channel.js]: ...then we don't need to check here

File: node/v2/handler.js [hunk 4]
```diff
@@ -180,6 +181,14 @@ TChannelV2Handler.prototype.handleCallRequest = function handleCallRequest(reqFr
     if (self.remoteHostPort === null) {
         return callback(errors.TChannelCallReqBeforeInitReqError());
     }
+
+    if (!reqFrame.body ||
+        !reqFrame.body.headers ||
+        !reqFrame.body.headers.as
+    ) {
+        return callback(errors.AsHeaderRequired());
```
Comment (jcorbin) [inline node/v2/handler.js]: Not callback(err), that's for low level "should close the socket" type errors.

In this case, this should be an operational error resulting in a `"BadRequest", "missing as header"` error frame to the caller.

**Comment by jcorbin** [comment]
:+1: let's:
- fix the handler guard so that it correctly sends error frame as noted
- then make that opt-outable by a channel option that turns it only into a local warn/error log (maybe stat)
- we had verbale buy in from the python side to make this required, so maybe even the option isn't necessary

**Comment by Raynos** [comment]
@jcorbin @kriskowal ready for review.

**Comment by kriskowal** [comment]
:ok:

From my regular editorial column, “0. naming things”:

> forceAs reads, “regardless of whether `as` is provided, use this one”, whereas requireAs, “ensure that `as` has been provided”.

**Comment by Raynos** [comment]
@kriskowal renamed.

**Comment by kriskowal** [comment]
:+1:

File: node/v2/handler.js [hunk 5]
```diff
@@ -181,7 +184,29 @@ TChannelV2Handler.prototype.handleCallRequest = function handleCallRequest(reqFr
...
+            var err = errors.AsHeaderRequired({
+                frame: 'request'
+            });
+            self.sendErrorFrame(
+                req.res, 'ProtocolError', err.message
+            );
+            return callback();
+        } else {
+            self.logger.error('Expected "as" header for incoming req', {
+                arg1: String(reqFrame.body.args[0]),
```
Comment (jcorbin) [inline node/v2/handler.js]: more caller info? remote addr?

File: node/v2/handler.js [hunk 6]
```diff
@@ -212,6 +237,23 @@ TChannelV2Handler.prototype.handleCallResponse = function handleCallResponse(res
...
+        !resFrame.body.headers.as
+    ) {
+        if (self.requireAs) {
+            var err = errors.AsHeaderRequired({
+                frame: 'response'
+            });
+            return callback(err);
+        } else {
+            self.logger.error('Expected "as" for incoming response', {
+                code: resFrame.body.code
```
Comment (jcorbin) [inline node/v2/handler.js]: moar context plox

File: node/v2/handler.js [hunk 7]
```diff
@@ -379,6 +421,17 @@ TChannelV2Handler.prototype.sendCallRequestFrame = function sendCallRequestFrame
     var reqBody = new v2.CallRequest(
         flags, req.ttl, req.tracing, req.serviceName, req.headers,
         req.checksum.type, args);
+
+    if (self.requireAs) {
+        assert(req.headers && req.headers.as,
+            'Expected the "as" transport header to be set for request');
+    } else {
+        self.logger.error('Expected "as" header to be set for request', {
+            arg1: String(args[0]),
```
Comment (jcorbin) [inline node/v2/handler.js]: y u no moar context

File: node/v2/handler.js [hunk 8]
```diff
@@ -392,6 +445,16 @@ TChannelV2Handler.prototype.sendCallResponseFrame = function sendCallResponseFra
     var resBody = new v2.CallResponse(
         flags, code, res.tracing, res.headers,
         res.checksum.type, args);
+
+    if (self.requireAs) {
+        assert(res.headers && res.headers.as,
+            'Expected the "as" transport header to be set for response');
+    } else {
+        self.logger.error('Expected "as" header to be set for response', {
+            code: code
```
Comment (jcorbin) [inline node/v2/handler.js]: con: text

**Comment by jcorbin** [comment]
:+1: good enough, could be better

**Comment by Raynos** [comment]
Added the context.

**Pull request merged**